### PR TITLE
Introduce `RenderingTarget` type and `CombinedProps` pattern

### PR DESCRIPTION
**Pull request by OllysCoding**
Based on the RFC #7174, this PR creates the `RenderingTarget` type and `CombinedProps` pattern.

**DCR & AR are becoming one** - specifically we are adding the capability for DCR to render articles for the app.

We will continue to use the Frontend model for both web & apps articles, and generally web and apps articles will be the same. However, there are some differences in requirements, mainly based on the technical impact of the target: 
- Apps uses native implementations of some features via bridget
- Apps requires dark mode support
- & others

To support these differences, this PR introduces the idea of a `RenderingTarget`, an enum which allows a component or function to know whether the content being rendered will be shown either on the web or the apps.

Additionally, to support differing requirements of data - a rare but nonetheless necessary occurrence - this PR introduces the `CombinedProps` type & pattern which allows the passing of 'web only' or 'apps only' props, ensuring our type safety is not diluted when rendering for different targets.

Most of the additional details & reasoning are outlined in the RFC: #7174

An important part of this PR is the JSDoc documentation, so please provide feedback of the wording can be improved or better formatted!

### CODE REVIEW

File: dotcom-rendering/.eslintrc.js [hunk 1]
```diff
@@ -120,7 +120,8 @@ module.exports = {
 		'no-empty-pattern': 'error',
 		'no-fallthrough': 'off', // We use 'noFallthroughCasesInSwitch' in tsconfig.json as this respects types
 		'no-param-reassign': 'error',
-		'no-shadow': 'warn',
+		'no-shadow': 'off', // We use the typescript-eslint version as eslint false positives on enums
+		'@typescript-eslint/no-shadow': ['error'],
```
Comment (OllysCoding) [inline dotcom-rendering/.eslintrc.js:124]: Without making this change, we'd getting a `variable is already declared in upper scope` warning from ESLint, despite that not being the case. Using the typescript version of this rule fixes this problem

File: dotcom-rendering/src/web/components/Callout/FormField.tsx [hunk 2]
```diff
@@ -125,10 +125,10 @@ export const FormField = ({
 						},
 					]
 						.concat(formField.options)
-						.map(({ value, label }) => {
+						.map(({ value, label: formLabel }) => {
```
Comment (OllysCoding) [inline dotcom-rendering/src/web/components/Callout/FormField.tsx:128]: Fix a 'no shadow' rule

File: dotcom-rendering/src/web/components/Carousel.importable.tsx [hunk 3]
```diff
@@ -505,9 +505,7 @@ export const Carousel = ({ heading, trails, onwardsSource, format }: Props) => {
 
 		const scrolled = current.scrollLeft + offset;
 
-		const nextOffset = offsets
-			.reverse()
-			.find((offset) => offset < scrolled);
+		const nextOffset = offsets.reverse().find((o) => o < scrolled);
```
Comment (OllysCoding) [inline dotcom-rendering/src/web/components/Carousel.importable.tsx:508]: Fix a 'no shadow' error

**Review by JamieB-gu — APPROVED** [review]
Nice! Very thorough documentation.

**Review by bryophyta — APPROVED** [review]
This looks good to me! I've just added one small comment about documentation 🙂

**Review by georgeblahblah — APPROVED** [review]
Looks good to me! I really appreciated the clear explanations

File: dotcom-rendering/src/types/renderingTarget.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ * might be differences in the requirements of each target.
+ *
+ * Targets:
+ * - Web 	=> A full web browser, such as chrome or safari on a desktop computer, laptop or phone.
+ * - Apps 	=> A webview rendered within the Android or iOS live apps
+ */
+export enum RenderingTarget {
+	Web,
+	Apps,
+}
```
Comment (mxdvl) [inline dotcom-rendering/src/types/renderingTarget.ts:15]: As [I mentioned in the RFC](https://github.com/guardian/dotcom-rendering/issues/7256#issuecomment-1462488634), I would recommend avoiding TypeScript enums, as they “are one of the few features TypeScript has which is not a type-level extension of JavaScript”[^1].

To prove my point, what do you expect the value of `RenderingTarget[0]` to be? What about `RenderingTarget.Web === 0`?

[^1]: https://www.typescriptlang.org/docs/handbook/enums.html

File: dotcom-rendering/src/web/components/Carousel.importable.tsx (hunk 3 above)
Comment (mxdvl) [reply dotcom-rendering/src/web/components/Carousel.importable.tsx:508]: Could also drop the `reverse` with [`Array.prototype.findLast` ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findLast)

**Review by mxdvl — COMMENTED** [review]
This looks like it bring a lot if type safety and is well documented: great!

Should the `no-shadow` change be extracted out of this PR?

Is the complexity of this wrapper type worth it compared to a naive implementation?

```ts
interface CommonProps {
  value: string;
  target: RenderingTarget; // enum or string union
}

interface AppProps extends CommonProps {
  target: "Apps";
  apiUrl: string;
}

interface WebProps extends CommonProps {
  target: "Web";
}

cont Component = (props: AppProps | WebProps) => <>{/* … */}</>
```

File: dotcom-rendering/src/types/renderingTarget.ts (hunk 4 above)
Comment (OllysCoding) [reply dotcom-rendering/src/types/renderingTarget.ts:15]: Would your suggestion be to use a string union? What's the best alternative

Comment (mxdvl) [reply dotcom-rendering/src/types/renderingTarget.ts:15]: Yes, string unions are a lot less confusing. If you need to do custom type-guard, the pattern I shared in the RFC works well:

> ```ts
> // If you define a array `as const`…
> const renderingTarget = [
>     "Web",
>     "Apps",
>     "Editions",
> ] as const;
> 
> // … you can infer the type from it…
> export type RenderingTarget = renderingTarget[number];
> 
> // … and create a custom type guard, if you so desire!
> export isRenderingTarget = (target: string): target is RenderingTarget =>
> 	renderingTarget.map(String).includes(target)
> ```

**Review by JamieB-gu — APPROVED** [review]

File: dotcom-rendering/.eslintrc.js (hunk 1 above)
Comment (mxdvl) [reply dotcom-rendering/.eslintrc.js:124]: Was this needed even without the use of TypeScript `enum`s?

**Pull request merged**